richtextfx/src/main/java/org/fxmisc/richtext/GenericStyledArea.java

```diff
@@ -0,0 +1,1371 @@
+            popup.setAnchorX(anchor.getX());
+            popup.setAnchorY(anchor.getY());
+        });
+    }
+
+    private Optional<Bounds> getCaretBoundsOnScreen() {
+        return virtualFlow.getCellIfVisible(getCurrentParagraph())
+                .map(c -> c.getNode().getCaretBoundsOnScreen());
+    }
+
+    private Optional<Bounds> getSelectionBoundsOnScreen() {
+        IndexRange selection = getSelection();
+        if(selection.getLength() == 0) {
+            return getCaretBoundsOnScreen();
+        }
+
+        Bounds[] bounds = virtualFlow.visibleCells().stream()
+                .map(c -> c.getNode().getSelectionBoundsOnScreen())
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .toArray(Bounds[]::new);
+
+        if(bounds.length == 0) {
+            return Optional.empty();
+        }
+        double minX = Stream.of(bounds).mapToDouble(Bounds::getMinX).min().getAsDouble();
+        double maxX = Stream.of(bounds).mapToDouble(Bounds::getMaxX).max().getAsDouble();
+        double minY = Stream.of(bounds).mapToDouble(Bounds::getMinY).min().getAsDouble();
+        double maxY = Stream.of(bounds).mapToDouble(Bounds::getMaxY).max().getAsDouble();
+        return Optional.of(new BoundingBox(minX, minY, maxX-minX, maxY-minY));
+    }
+
+    private <T> void subscribeTo(EventStream<T> src, Consumer<T> cOnsumer) {
+        manageSubscription(src.subscribe(cOnsumer));
+    }
+
+    private void manageSubscription(Subscription subscription) {
+        subscriptions = subscriptions.and(subscription);
+    }
+
+    private void manageBinding(Binding<?> binding) {
+        subscriptions = subscriptions.and(binding::dispose);
+    }
+
+    private static Bounds extendLeft(Bounds b, double w) {
+        if(w == 0) {
+            return b;
+        } else {
+            return new BoundingBox(
+                    b.getMinX() - w, b.getMinY(),
+                    b.getWidth() + w, b.getHeight());
+        }
+    }
+
+    void clearTargetCaretOffset() {
+        targetCaretOffset = Optional.empty();
+    }
+
+    ParagraphBox.CaretOffsetX getTargetCaretOffset() {
+        if(!targetCaretOffset.isPresent())
+            targetCaretOffset = Optional.of(getCaretOffsetX());
+        return targetCaretOffset.get();
+    }
+
+    private static EventStream<Boolean> booleanPulse(javafx.util.Duration javafxDuration, EventStream<?> restartImpulse) {
+        Duration duration = Duration.ofMillis(Math.round(javafxDuration.toMillis()));
+        EventStream<?> ticks = EventStreams.restartableTicks(duration, restartImpulse);
+        return StateMachine.init(false)
+                .on(restartImpulse.withDefaultEvent(null)).transition((state, impulse) -> true)
+                .on(ticks).transition((state, tick) -> !state)
+                .toStateStream();
+    }
+
+
+    public static <SEG, S> Node createStyledTextNode(SEG seg, SegmentOps<SEG, S> segOps, BiConsumer<? super TextExt, S> applyStyle) {
```

Comment:
This methods should be specialized to `StyledText`, since it doesn't work for and shouldn't be used with any other segments, right?